hs-bindgen/test-th/HsBindgen/TestTH/Examples.hs

```diff
@@ -0,0 +1,58 @@
+{-# OPTIONS_GHC -Wno-incomplete-uni-patterns #-}
```

Comment:
blargh. Typed interface, yeah, but with partial bindings. Safe, but not safe.

Reply:
This is the eternal tradeoff: how much information do we want to include in types?

Reply:
I'd argue we do, as we won't get "partial pattern match" exceptions.

I really want the compiler to be able to help me, to construct and consume valid structures.

Reply:
Yes, true, there is a benefit to it, I agree. Perhaps I should rephrase my question: is that benefit worth the cost of tracking bound vars at the type level? We'd have to have type-level naturals, type-level addition (so that we can track how many variables we bind when we match on a struct), we'd have to record at the type-level what the size of structs are, we'd then have to introduce those as existentials when we do the translation from C, ... It's possible, but quite a bit of machinery. As it stands, we get _some_ type safety (more than in a first-order representation), and I'd argue some convenience also, without too much cost. 

But if you think we can track the vars at the type-level without too much cost, I'd love to see. I do agree that it would be nicer.

Reply:
Thinking about this some more: one benefit actually of having the size available at the type level is that the `Show` instances also become easier (it's a bit tricky to `Show` open terms right now; possible, but a bit ugly). Maybe I will experiment with a richer representation that does in fact track bound variables at the type-level; the case that case me most concerned is pattern matching on structs, but perhaps we can avoid type-level addition by instead having a type-level list instead of a single `Nat`, so that we can "index" variables. Something like

```haskell
'[ 'Z, 'Z, 'Z ]
```

would then just to equivalent to 3 (three bound variables), but if we have

```haskell
'[ 'S 'S 'Z, 'Z, 'Z ]
```

then perhaps we have three bound variables, `x`, `y`, and `z.0`, `z.1`, `z.2`. Maybe this way the overhead isn't too large. I'll experiment a bit with this.

Reply:
Experimented a tiny bit, I think it might be workable, having terms indexed by `[Nat]` rather than `Nat`, for available "subscripted" variables; however, if we do only that, we're back to essentially dealing with DeBruijn indices when constructing terms, which is unpleasant. Perhaps that we can restore the higher-order interface using something like [I am not a number, I am a classy hack](https://mazzo.li/epilogue/index.html%3Fp=773.html) but I am out of time for now. To be continued.

Reply:
Still not convinced it's worth it though.

Reply:
You overly complicated this. 

```haskell
variableLambda :: Size n -> (Vec n Bound -> Term) -> Term
```

or

```haskell
variableLambda :: Traversable t
  => t Bound -- ^ names
  -> (t Bound -> Term) -- ^ body
  -> Term
```

which also couples the arity of binder to the body. (The consumer cannot do anything else with names `t Bound` other than `traverse` - so they cannot come up with wrong arity container to pass to the body-function).

no need to track the arity outside of a term.

Reply:
Ah! Good idea. Ok, will do. I like the `Vec` approach better, it allows to continue to write clean-looking manual examples; I think the `Traversable` thing will still be tricky (how do you write somehing like

```haskell
    , storablePoke = Hs.Lambda $ \ptr -> Hs.Lambda $ \x ->
        Hs.ElimStruct myStruct x $ \[f1, f2] -> Hs.Seq [
            Hs.PokeByteOff ptr 0 f1
          , Hs.PokeByteOff ptr 4 f2
          ]
```

in a type-safe way if all you have is `Traversable`? Much easier with the `Vec`-like approach. Will do (perhaps `NP` instead of `Vec`).

Reply:
You can instantiate `t` with some `Vec`, but you don't need to.